clippy_lints/src/legacy_numeric_constants.rs

```diff
@@ -0,0 +1,227 @@
+use rustc_middle::{hir::nested_filter::OnlyBodies, lint::in_external_macro};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, Span, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `<integer>::max_value()`, `std::<integer>::MAX`,
+    /// `std::<float>::EPSILON`, etc.
+    ///
+    /// ### Why is this bad?
+    /// All of these have been superceded by the associated constants on their respective types,
+    /// such as `i128::MAX`. These legacy constants may be deprecated in a future version of rust.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let eps = std::f32::EPSILON;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let eps = f32::EPSILON;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub LEGACY_NUMERIC_CONSTANTS,
+    style,
+    "checks for usage of legacy std numeric constants and methods"
+}
+pub struct LegacyNumericConstants {
+    msrv: Msrv,
+    use_stmts: FxHashMap<Symbol, Vec<Span>>,
+    glob_use_stmts: Vec<Span>,
+}
+
+impl LegacyNumericConstants {
+    #[must_use]
+    pub fn new(msrv: Msrv) -> Self {
+        Self {
+            msrv,
+            use_stmts: FxHashMap::default(),
+            glob_use_stmts: vec![],
+        }
+    }
+}
+
+impl_lint_pass!(LegacyNumericConstants => [LEGACY_NUMERIC_CONSTANTS]);
+
+impl<'tcx> LateLintPass<'tcx> for LegacyNumericConstants {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        let Self {
+            msrv: _,
+            use_stmts,
+            glob_use_stmts,
+        } = self;
+
+        if !item.span.is_dummy() && let ItemKind::Use(path, kind) = item.kind {
+            match kind {
+                UseKind::Single => {
+                    for res in &path.res {
+                        if let Some(def_id) = res.opt_def_id()
+                            && let Some(module_name) = is_path_in_integral_module(cx, def_id)
+                            && let _ = use_stmts.insert(module_name, vec![])
+                            && let Some(use_stmts) = use_stmts.get_mut(&module_name)
+                        {
+                            use_stmts.push(item.span);
+                        }
+                    }
+                },
+                UseKind::Glob => glob_use_stmts.push(item.span),
+                UseKind::ListStem => {},
+            }
+        }
+    }
+
+    fn check_crate_post(&mut self, cx: &LateContext<'tcx>) {
+        let Self {
+            msrv,
+            use_stmts,
+            glob_use_stmts,
+        } = self;
+
+        let mut v = V {
+            cx,
+            msrv,
+            use_stmts,
+            glob_use_stmts,
+        };
+        cx.tcx.hir().visit_all_item_likes_in_crate(&mut v);
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+struct V<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    msrv: &'a Msrv,
+    use_stmts: &'a FxHashMap<Symbol, Vec<Span>>,
+    glob_use_stmts: &'a Vec<Span>,
+}
```

Comment:
I'm not sure, if visiting everything again with a Visitor is the best thing to do. I'm guessing, you're doing it this way, to catch cases where the `use std::primitive::u32` is after an item/body?

I can't remember if this has been discussed as part of the previous conversion. But I would guess that just linting on `use std::primitive::u32` should be fine. There might be some uncool cases like `use std::primitive::u32::MAX` but even then, I'd say it's enough to lint that one and state that all uses of `MAX` have to be replaced with `u32::MAX`. Walking all items again will probably have a performance impact and is weird when it comes to msrv tracking and lint emission.

Reply:
Yeah we need to do 2 passes to have that work unfortunately, we could lint the `use` specifically if it exists but that might be a bit uglier than just linting the usage

Reply:
Linting the `use` seems fine to me since that's the part we want them to remove

Reply:
But we can't really know whether they're using it for `std::f32::MAX` or `std::f32::consts`. Something like
```rust
use std::f32;
use f32::consts;
```
Would be a FP (but not something people use regardless)

Reply:
I think this would be an acceptable FP, I can't imagine people really using this. :)

Reply:
There's quite a few cases of it. https://github.com/search?q=%2Fuse+f32%3A%3Aconsts%7Cuse+f64%3A%3Aconsts%2F&type=code

Reply:
How about this:

- lint `use std::<int>` directly, but not `std::<float>`
- do not lint `<int>::<CONST>` even if it resolves to `std::<int>::<CONST>`, because it's covered by the previous
- lint all other cases of `std::<int>::<CONST>`: `MAX`, `std::i32::MIN`, etc
- lint all cases of `std::<float>::<CONST>`
- provide an unspanned "remove the import" help or note for cases where `<float>::<CONST>` resolves to `std::<float>::<CONST>`

Reply:
I think this is a good idea and should also simplify the implementation :)

Reply:
I don't particularly agree, it's inconsistent with the behavior of `deprecated` which lints on the usage itself which is generally a lot clearer. In the case of `std::<int>` though we should also lint the `use` itself as that's also quasi-deprecated but regardless the usage itself should be linted

Seeing `usage of legacy numeric constants` on `use std::u32::MAX` will likely confuse the programmer if removing that suddenly causes compilation to fail, and they may just silence the lint as a result (adding a single `allow` is easier than manually changing every `MAX` to `u32::MAX`)

This would also make it non-`MachineApplicable` in the case of `std::f32::MAX` and friends, which I don't think is worth the increased simplicity...

Reply:
I think you may have misunderstood the behavior I proposed.

> Seeing `usage of legacy numeric constants` on `use std::u32::MAX` will likely confuse the programmer if removing that suddenly causes compilation to fail, and they may just silence the lint as a result (adding a single `allow` is easier than manually changing every `MAX` to `u32::MAX`)

Under the behavior I suggested, `use std::u32::MAX;` would NOT trigger the lint. `use std::u32;` would.

> This would also make it non-`MachineApplicable` in the case of `std::f32::MAX` and friends, which I don't think is worth the increased simplicity...

The behavior for `std::f32::MAX` would not change at all under my proposal.

Reply:
I see.
But I still don't think we should go with this. In the case of `f32::MAX` or `f64::MAX` it'll result in a much less helpful message, just "remove the import" which is ambiguous; which import? It looks identical to `f32::MAX`, just saying there's some import for `std::f32` *somewhere* isn't all too helpful.

What this simplifies is removing the `use` collection; however, the rest of it is only made more complicated. If the "remove the import" portion is improved, nothing has been simplified.

> But the pattern `i32::MAX` is not the problem, it's actually the prior `use std::i32;` that's the problem because it was made redundant.

And in the cases where they're identical, It should still lint the usage but not give a suggestion, only listing the potentially bad imports. Then, since `std::i32` is quasi-deprecated, that can be linted as well (we could also omit the bad imports in this case, since it'll be linted). In the case of `f32::MAX` it'd always show the bad imports.

Reply:
> But I still don't think we should go with this. In the case of `f32::MAX` or `f64::MAX` it'll result in a much less helpful message, just "remove the import" which is ambiguous; which import? It looks identical to `f32::MAX`, just saying there's some import for `std::f32` *somewhere* isn't all too helpful.

I think you're overstating how hard it is to find and remove an import. And in the case of floats, it will commonly require more work anyways to resolve the `::consts` issue. Maybe we just shouldn't ever lint `f32::MAX`.

> What this simplifies is removing the `use` collection; however, the rest of it is only made more complicated. If the "remove the import" portion is improved, nothing has been simplified.
> 
> And in the cases where they're identical, It should still lint the usage but not give a suggestion, only listing the potentially bad imports. Then, since `std::i32` is quasi-deprecated, that can be linted as well (we could also omit the bad imports in this case, since it'll be linted). In the case of `f32::MAX` it'd always show the bad imports.

In my opinion, linting `i32::MAX` is purely noise. There's nothing to change with that code. All that needs to be done is remove the redundant import. With current behavior, stderr is longer and more repetitive.

It feels like throwing a warning when someone uses `.try_into()` because they don't need to explicitly import the `TryInto` trait, since it's in the prelude.

Here's how I see it. The current behavior is slower, more complicated, and more repetitive for a slightly more useful warning message. I don't think the tradeoffs are worth it.